File: Week_05/G20190343020039/LeetCode_91_0039.js
/**
 * @param {string} s
 * @return {number}
 */
var numDecodings = function(s) {
  if (s === null || s.length === 0) { return 0; }
  let dp = new Array(s.length).fill(0);
  dp[0] = s.charAt(0) !== '0' ? 1 : 0;
  for (let i = 1; i < s.length; i += 1) {
    const first = +s.charAt(i);
    const second = +s.substring(i-1, i+1);
    if (first >= 1 && first <= 9) {
      dp[i] += dp[i-1];
    }
    if (second >= 10 && second <= 26) {
      dp[i] += i >= 2 ? dp[i-2] : 1;
    }
  }
  return dp[s.length-1];
};

File: Week_08/G20190343020039/LeetCode_91_0039.py
class Solution:
  # dp[i] = dp[i-1] + dp[i-2]
  def numDecodings(self, s: str) -> int:
    if len(s) == 0: return 0
    dp = [0] * len(s)
    if s[0] != '0':
      dp[0] = 1
    for i in range(1, len(s)):
      first = int(s[i])
      second = int(s[i-1:i+1])
      if first >= 1 and first <= 9:
        dp[i] += dp[i-1]
      if second >= 10 and second <= 26:
        if i >= 2:
          dp[i] += dp[i-2]
        else:
          dp[i] += 1
    return dp[-1]

